src/hosted-buttons/utils.js

```diff
@@ -0,0 +1,157 @@
+/* @flow */
+
+import { request, memoize, popup } from "@krakenjs/belter/src";
+import { getSDKHost, getClientID } from "@paypal/sdk-client/src";
+
+import { DEFAULT_POPUP_SIZE } from "../zoid/checkout";
+
+import type {
+  ButtonVariables,
+  CreateAccessToken,
+  CreateOrder,
+  GetCallbackProps,
+  HostedButtonDetailsParams,
+  OnApprove,
+  RenderForm,
+} from "./types";
+
+const entryPoint = "SDK";
+const baseUrl = `https://${getSDKHost()}`;
+const apiUrl = baseUrl.replace("www", "api");
+
+const getHeaders = (accessToken?: string) => ({
+  ...(accessToken && { Authorization: `Bearer ${accessToken}` }),
+  "Content-Type": "application/json",
+  "PayPal-Entry-Point": entryPoint,
+});
+
+export const createAccessToken: CreateAccessToken = memoize<CreateAccessToken>(
+  (clientId) => {
+    return request({
+      url: `${apiUrl}/v1/oauth2/token`,
+      method: "POST",
+      body: "grant_type=client_credentials",
+      headers: {
+        Authorization: `Basic ${btoa(clientId)}`,
+        "Content-Type": "application/json",
+      },
+    }).then((response) => response.body.access_token);
+  }
+);
+
+const getButtonVariable = (variables: ButtonVariables, key: string): string =>
+  variables?.find((variable) => variable.name === key)?.value ?? "";
+
+const getFundingSource = (paymentSource) => {
+  if (paymentSource === "credit") {
+    return `CARD`;
+  }
+  return paymentSource.toUpperCase();
+};
+
+export const getHostedButtonDetails: HostedButtonDetailsParams = ({
+  hostedButtonId,
+}) => {
+  return request({
+    url: `${baseUrl}/ncp/api/form-fields/${hostedButtonId}`,
+    headers: getHeaders(),
+  }).then(({ body }) => {
+    const variables = body.button_details.link_variables;
+    return {
+      style: {
+        layout: getButtonVariable(variables, "layout"),
+        shape: getButtonVariable(variables, "shape"),
+        color: getButtonVariable(variables, "color"),
+        label: getButtonVariable(variables, "button_text"),
+      },
+      html: body.html,
+      htmlScript: body.html_script,
+    };
+  });
+};
+
+/**
+ * Attaches form fields (html) to the given selector, and
+ * initializes window.__pp_form_fields (htmlScript).
+ */
+export const renderForm: RenderForm = ({
+  hostedButtonId,
+  html,
+  htmlScript,
+  selector,
+}) => {
+  const elm =
+    typeof selector === "string" ? document.querySelector(selector) : selector;
+  if (elm) {
+    elm.innerHTML = html + htmlScript;
+    const newScriptEl = document.createElement("script");
+    const oldScriptEl = elm.querySelector("script");
```

Comment:
Will there only ever be one script per element? I'm assuming so, but thought I'd doublecheck.

Reply:
that is right! this is only because it is sent as one property in the response here - https://github.com/paypal/paypal-checkout-components/pull/2276/files#diff-dc0ab5efddb0f4c5dc8c699ccafce1daa87906e1b839480c5be57048cee5482bR88